pkg/controller/machine/machine.go

```diff
@@ -420,28 +428,50 @@ func (c *Controller) cleanupMachineAfterDeletion(machine *machinev1alpha1.Machin
 	return nil
 }
 
-// deleteMachineAndProviderInstance makes sure that an instance has gone in a series of steps.
-func (c *Controller) deleteMachineAndProviderInstance(prov cloud.Provider, machine *machinev1alpha1.Machine) error {
+// deleteMachine makes sure that an instance has gone in a series of steps.
+func (c *Controller) deleteMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
 	if err := c.deleteProviderInstance(prov, machine); err != nil {
-		message := fmt.Sprintf("%v. Please manually delete finalizers from the machine object.", err)
-		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete machine: %v", err)
-		return c.updateMachineErrorIfTerminalError(machine, machinev1alpha1.DeleteMachineError, message, err, "failed to delete machine at cloudprovider")
+		message := fmt.Sprintf("%v. Please manually delete %s finalizer from the machine object.", err, finalizerDeleteInstance)
+		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete instance: %v", err)
+		return c.updateMachineErrorIfTerminalError(machine, common.DeleteMachineError, message, err, "failed to delete machine at cloudprovider")
+	}
+	if err := c.deleteNodeForMachine(machine); err != nil {
+		message := fmt.Sprintf("%v. Please manually delete %s finalizer from the machine object.", err, finalizerDeleteNode)
+		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete node: %v", err)
+		return errors.New(message)
 	}
 	return c.cleanupMachineAfterDeletion(machine)
 }
 
-func (c *Controller) ensureInstanceExistsForMachine(prov cloud.Provider, machine *machinev1alpha1.Machine, userdataProvider userdata.Provider, providerConfig *providerconfig.Config) error {
+func (c *Controller) deleteNodeForMachine(machine *clusterv1alpha1.Machine) error {
+	nodesList, err := c.nodesLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list nodes: %v", err)
+	}
+
+	for _, node := range nodesList {
+		if ownerUID, exists := node.GetLabels()[NodeOwnerLabelName]; exists {
+			if ownerUID == string(machine.UID) {
+				return c.kubeClient.CoreV1().Nodes().Delete(node.Name, nil)
```

Comment:
As i just found out about it: `c.nodesLister.ListWithPredicate` allows to pass a custom filter function.
So only objects, which apply to the filter function are being returned.

Just wanted to mention it, not requesting a change here.